#32T check fully reviewed before approved based

nmadruga opened the pull request:
Take 2 - Had problems with conflicts when rebasing to `develop`

[Back-end PR](https://github.com/openmsupply/conforma-server/pull/921)

Part 2 - Fixes https://github.com/openmsupply/conforma-templates/issues/32 & fixes #1180 

~~[Back-end Base PR](880) - Part 1~~

### Testing
Use `dev/snapshots/AA_Testing_Fully_reviewed_submission` to test with one partial review already in place. 

1. Log in as this fake user `reviewer`/`ARMED1234`. 
2. There are 2 applications for ProdRegistration in to be reviewed. 
I have already assigned so 1 application ("Test Medicamento") can be completed (as fully assigned) and the next one should show the warning to the user, even though the review can be submitted, but the incrementStage should not meet conditions. 
The setup is that applications:
 - "Test Medicamento" only has Section 1 elements submitted => If Reviewer approves it should go to Stage 2
 - "Test Medicamento 2" has one extra element in Section 2 submitted => Reviewer has been assigned to Section 1 only. When submitting the review, it should show a warning and allow to submit but **keep the application in Stage 1**.
 - You can try after that assign the other Section and try to submit... I haven't tried this step.

### Changes
- Warning when user submits Review/Consolidation that doesn't have all assigned and submitted questions (reviewable)
- The warning shows, but user can still submit - It should allow submitting the partial reviewsin order to allow for different users to review different sections. The last one submitting will trigger the action to Increment Stage correctly.
- I have also to update the **CORE-ACTION** _IncrementStage_ to be checking for fully approved elements... That's what prevent it moving to next stage when not fully reviewed. 
- **This hasn't been added to core_actions snapshot - I'll do before this PR is merged** 

### Pics
Message displayed when trying to submit (partial - not fully approved)
![Screen Shot 2022-10-17 at 3 58 13 PM](https://user-images.githubusercontent.com/16461988/196079790-14b30670-e19e-4f2b-93db-122ffa01bafb.png)

Checking graphql functions:
*Check should work on Application 1, since*:
![Screen Shot 2022-10-17 at 3 13 14 PM](https://user-images.githubusercontent.com/16461988/196079886-4be82c8d-2fe7-498a-acc3-2612c8a8b86d.png)
![Screen Shot 2022-10-17 at 3 12 34 PM](https://user-images.githubusercontent.com/16461988/196079888-753917fb-a4a1-447a-bc71-f25cc9c7fe35.png)

*Check should fail on Application 2, since*:
![Screen Shot 2022-10-17 at 3 11 41 PM](https://user-images.githubusercontent.com/16461988/196080096-1161568d-a696-4cda-82e9-bbc5117c3130.png)
![Screen Shot 2022-10-17 at 3 11 54 PM](https://user-images.githubusercontent.com/16461988/196079895-488cc351-692e-47a9-b169-1f53cf3db2c8.png)

nmadruga commented on src/utils/defaultLanguageStrings.ts:
```diff
@@ -329,6 +329,8 @@ export default {
   REVIEW_SUBMISSION_FAIL: 'Not all sections have been reviewed',
   REVIEW_SUBMISSION_ERROR: 'Something wrong with server - review mutation failed',
+  REVIEW_SUBMISSION_INCOMPLETE_MESSAGE:
+    'You are about to submit your review. Some sections have not be completed by others, which means your decision is not changing the outcome - if aplicable. Are you sure you wish to proceed?',
```
Oh, I only meant - In case approving our change the Outcome of the application. But if you think that's confusing I can rephrase. I'll try something better now.

nmadruga replied:
Done in [bf41f47](https://github.com/openmsupply/conforma-web-app/pull/1391/commits/bf41f4735257476f5047c04b984e9d3e2c7706b1)

Please have a look see if it's a bit better now

CarlosNZ replied:
Cool, maybe just a little more concise (and small grammar tweaks):
"You are about to submit your review, although your decision will not change the outcome yet as some sections are not yet completed by other reviewers. Are you sure you wish to proceed?"

nmadruga replied:
Done [0f11154](https://github.com/openmsupply/conforma-web-app/pull/1391/commits/0f1115484133adf7f3974c0730518dd66ddcd808)

nmadruga commented on src/utils/hooks/useSubmitReview.ts:
```diff
@@ -33,8 +33,6 @@ const useSubmitReview: UseSubmitReview = (reviewId, reload) => {
 
     const reviewResponsesPatches = reviewResponses.map(({ thisReviewLatestResponse }) => ({
       patch: {
-        decision: thisReviewLatestResponse?.decision,
-        comment: thisReviewLatestResponse?.comment,
```
You have asked this before. But that's on an older closed PR (remade due to trying to keep it alive)

The explanation is here: https://github.com/openmsupply/conforma-web-app/pull/1357#discussion_r996491751
> It isn't required now because we already do that for each review when the reviewer clicks on the "Submit" review button or on the "Agree with all". There was a problem with this code now which would overwrite the decision with "No decision" option.

CarlosNZ replied:
Okay, I didn't remember seeing the reply the first time. Thanks.

nmadruga commented:
> > * "Test Medicamento" only has Section 1 elements submitted => If Reviewer approves it should go to Stage 2
> 
> This didn't happen -- stayed in stage Triagem. (Action_queue says CONDITION NOT MET)

Ohhh, I have seen this one happening before. But I thought it would work now. Time to use the super duper Triggers testing tool : D

nmadruga commented:
Note to self: Remember to update message in Portuguese + improve title in Portuguese to "Enviar avaliação"

nmadruga commented:
> > > * "Test Medicamento" only has Section 1 elements submitted => If Reviewer approves it should go to Stage 2
> > 
> > 
> > This didn't happen -- stayed in stage Triagem. (Action_queue says CONDITION NOT MET)
> 
> Ohhh, I have seen this one happening before. But I thought it would work now. Time to use the super duper Triggers testing tool : D

FIXED I'm happy it was an easy fix. The problem was that I changed the Function `approved_assigned_questions_count` but didn't use CREATE OR REPLACE. So it was failing and trying to check only for already submitted. Please try again now:

- Update the back-end
- Reload a new snapshot `dev/snapshots/AA_Testing_Fully_reviewed_submission_4` and try again please.

The result after submitting the first application should be showing it moved to the next Stage:
<img width="1195" alt="Screen Shot 2022-10-19 at 8 51 29 PM" src="https://user-images.githubusercontent.com/16461988/196630838-0c11a85a-647f-46c2-b778-3d8ebb60d71a.png">

CarlosNZ commented:
> I have also to update the **CORE-ACTION** _IncrementStage_ to be checking for fully approved elements... That's what prevent it moving to next stage when not fully reviewed.

It's my understanding that this is only required because this is a special case for Angola in that it can advance to the next stage without being a last-level review, is that correct? In which case, maybe we shouldn't put it in core_templates and over-complicate it -- the normal condition should be enough, wouldn't it?( isLastLevel and decision = CONFORM)

CarlosNZ reviewed (approved):
Okay, I'm happy with this, all looks good.

My only issue is with the config, as I don't think we need that new back-end function (written on back-end PR), but this front-end stuff seems solid. 

Thanks a lot for figuring this all out.

nmadruga commented:
> Note to self: Remember to update message in Portuguese + improve title in Portuguese to "Enviar avaliação" And this text has grammar mistake: <img alt="Screen Shot 2022-10-19 at 8 51 48 PM" width="444" src="https://user-images.githubusercontent.com/16461988/196630046-268996f3-f41b-4dca-8f72-354a55efe3d2.png">

Done in latest snapshot

nmadruga commented:
> It's my understanding that this is only required because this is a special case for Angola in that it can advance to the next stage without being a last-level review, is that correct? 

Well, actually in the case of ANY consolidation we would still want to apply this check. As per issue #1180

> In which case, maybe we shouldn't put it in core_templates and over-complicate it -- the normal condition should be enough, wouldn't it?( isLastLevel and decision = CONFORM)

I think following the idea of having 2 "Sample templates" on our future version of Core_templates snapshot - to be used also for the test-suite, the one with Consolidation should use a similar check as well. And this should help with future work where we need to use this check to prevent one Consolidation to APPROVE before all lower level reviews have been submitted.

nmadruga commented:
> Okay, I'm happy with this, all looks good.
> 
> My only issue is with the config, as I don't think we need that new back-end function (written on back-end PR), but this front-end stuff seems solid.

Unfortunately this PR requires the Front-end PR in order to allow me to configure the snapshot correctly. So, I'll have another look on the Front-end to check your suggestion. But this PR will be pending here until we have an agreement

nmadruga commented:
Alright! Ready to merge this now. I will update the example `AA_Testing_Fully_reviewed_submission` in the Conforma-templates repo folder `dev/snapshots` in case you would like to retest this. I have tested for mentioned cases in the description already.

It will be pending Config changes in our latest snapshot to prevent the IncrementStage from happening as in the example snapshot.
